x/auth/middleware/branch_store.go

```diff
@@ -68,3 +98,113 @@ func branchStore(sdkCtx sdk.Context, tx tmtypes.Tx) (sdk.Context, sdk.CacheMulti
 
 	return sdkCtx.WithMultiStore(msCache), msCache
 }
+
+// anteBranchWrite is the tx.Handler that commits the state writes of a previously
+// created anteBranchBegin tx.Handler.
+type anteBranchWrite struct {
+	next          tx.Handler
+	originalMSKey sdk.ContextKey
+	cacheMSKey    sdk.ContextKey
+}
+
+// CheckTx implements tx.Handler.CheckTx method.
+// Do nothing during CheckTx.
+func (sh anteBranchWrite) CheckTx(ctx context.Context, req tx.Request, checkReq tx.RequestCheckTx) (tx.Response, tx.ResponseCheckTx, error) {
+	newCtx := anteWrite(ctx, sh.originalMSKey, sh.cacheMSKey)
+
+	return sh.next.CheckTx(newCtx, req, checkReq)
+}
+
+// DeliverTx implements tx.Handler.DeliverTx method.
+func (sh anteBranchWrite) DeliverTx(ctx context.Context, req tx.Request) (tx.Response, error) {
+	newCtx := anteWrite(ctx, sh.originalMSKey, sh.cacheMSKey)
+
+	return sh.next.DeliverTx(newCtx, req)
+}
+
+// SimulateTx implements tx.Handler.SimulateTx method.
+func (sh anteBranchWrite) SimulateTx(ctx context.Context, req tx.Request) (tx.Response, error) {
+	newCtx := anteWrite(ctx, sh.originalMSKey, sh.cacheMSKey)
+
+	return sh.next.SimulateTx(newCtx, req)
+}
+
+// branchAndRun creates a new Context based on the existing Context with a MultiStore branch
+// in case message processing fails.
+func anteWrite(ctx context.Context, originalMSKey, cacheMSKey sdk.ContextKey) context.Context {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	originalStore := sdkCtx.Value(originalMSKey).(sdk.MultiStore)
+	branchedStore := sdkCtx.Value(cacheMSKey).(sdk.CacheMultiStore)
+
+	if !sdkCtx.IsZero() {
+		// At this point, newCtx.MultiStore() is a store branch, or something else
+		// replaced by the AnteHandler. We want the original multistore.
+		//
+		// Also, in the case of the tx aborting, we need to track gas consumed via
+		// the instantiated gas meter in the AnteHandler, so we update the context
+		// prior to returning.
+		sdkCtx = sdkCtx.WithMultiStore(originalStore)
+	}
+
+	branchedStore.Write()
```

Comment:
I'm confused...where do we actually check for errors if say there is an error in any of the middleware ante-handlers?

Reply:
If any of the antehandlers fails, then this `anteBranchWrite` one is never called. Take a look at the middleware stack in `middleware.go`, `endAnteBranch` is below all the antehandlers.